xen/arch/x86/tpm.c

```diff
@@ -703,48 +771,159 @@ static uint32_t tpm20_pcr_extend(unsigned loc, uint32_t pcr_handle,
 }
 
-static uint32_t tpm_hash_extend20(unsigned loc, uint8_t *buf, unsigned size,
-                                  unsigned pcr, uint8_t *sha1_digest,
-                                  uint8_t *sha256_digest)
+static uint32_t tpm2_hash_extend(unsigned loc, uint8_t *buf, unsigned size,
+                                 unsigned pcr,
+                                 const struct tpm2_log_hashes *log_hashes)
 {
     uint32_t rc;
+    unsigned i;
+    unsigned j;
+    struct tpm2_log_hashes supported_hashes = {0};
 
     request_locality(loc);
 
-    if ( tpm20_has_sha1(loc) )
-        sha1_hash(buf, size, (uint32_t *)sha1_digest);
-    else
-        sha1_digest = NULL;
+    j = 0;
+    for ( i = 0; i < log_hashes->count; ++i ) {
+        const struct tpm2_log_hash *hash = &log_hashes->hashes[i];
+        if ( !tpm_supports_hash(loc, hash) ) {
+            printk(XENLOG_WARNING "Skipped hash unsupported by TPM: %d\n",
+                   hash->alg);
+            continue;
+        }
 
-    sha256_hash(buf, size, (uint32_t *)sha256_digest);
+        if ( hash->alg == TPM_ALG_SHA1 )
+            sha1_hash(buf, size, (uint32_t *)hash->data);
+        else if ( hash->alg == TPM_ALG_SHA256 )
+            sha256_hash(buf, size, (uint32_t *)hash->data);
+        else
+            /* create_log_event20() took take of initializing the digest. */;
+
+        if ( supported_hashes.count == MAX_HASH_COUNT ) {
+            printk(XENLOG_ERR "Hit hash count implementation limit: %d\n",
+                   MAX_HASH_COUNT);
+            return -1;
+        }
 
-    rc = tpm20_pcr_extend(loc, HR_PCR + pcr, sha1_digest, sha256_digest);
+        supported_hashes.hashes[supported_hashes.count] = *hash;
+        ++supported_hashes.count;
+    }
 
+    rc = tpm20_pcr_extend(loc, HR_PCR + pcr, &supported_hashes);
     relinquish_locality(loc);
 
     return rc;
 }
 
 #endif /* __EARLY_TPM__ */
 
+static struct heap_event_log_pointer_element2_1 *find_evt_log_ext_data(void)
+{
+    struct txt_sinit_mle_data *sinit_mle;
+    struct txt_ext_data_element *ext_data;
+
+    sinit_mle = txt_sinit_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+    ext_data = (void *)((uint8_t *)sinit_mle + sizeof(*sinit_mle));
+
+    /*
+     * Find TXT_HEAP_EXTDATA_TYPE_EVENT_LOG_POINTER2_1 which is necessary to
+     * know where to put the next entry.
+     */
+    while ( ext_data->type != TXT_HEAP_EXTDATA_TYPE_END ) {
+        if ( ext_data->type == TXT_HEAP_EXTDATA_TYPE_EVENT_LOG_POINTER2_1 )
+            break;
+        ext_data = (void *)&ext_data->data[ext_data->size];
+    }
+
+    if ( ext_data->type == TXT_HEAP_EXTDATA_TYPE_END )
+        return NULL;
+
+    return (void *)&ext_data->data[0];
+}
+
+static struct tpm2_log_hashes
+create_log_event20(struct tpm2_spec_id_event *evt_log, uint32_t evt_log_size,
+                   uint32_t pcr, uint32_t type, uint8_t *data,
+                   unsigned data_size)
+{
+    struct tpm2_log_hashes log_hashes = {0};
```

Comment:
I think it would be safer to allocate this struct on caller's side and pass pointer to it as an argument. It works for now, but if for any reason (e.g. different optimization settings) this function wouldn't be inlined, compiler may choose to return this as a pointer to structure on called function's stack, which may be overwritten.

Reply:
Compiler optimizations must not change behaviour of the code in any way, they definitely shouldn't completely break  correct code, otherwise it's a compiler bug.

Returning structs is always implemented by allocating object on the caller side and passing pointer to the callee.  When callee's body is known, copying from local variable to result can be elided by writing directly into the result, which is an optimization.  So I don't believe this requires any changes, returning structs by value works fine in C.

Reply:
You are correct, for some reason I thought that a pointer was returned instead of whole struct.

Just for completeness of discussion, structs can also be returned in registers "if they fit", but what that exactly means depends on compiler, calling convention and additional switches used. Nevertheless, this won't be an issue here, both because the structure is relatively big and because both caller and called function are in the same compile unit, so they always use the same way of passing data between each other.